TLS certificate object documentation and support for EC certificates (#24358)
nodejs/node: doc/api/tls.md

* `detailed` {boolean} Include the full certificate chain if `true`, otherwise
  include just the peer's certificate.
* Returns: {Object} A certificate object.

Returns an object representing the peer's certificate. If the peer does not
provide a certificate, an empty object will be returned. If the socket has been
destroyed, `null` will be returned.

If the full certificate chain was requested, each certificate will include an
`issuerCertificate` property containing an object representing its issuer's
certificate.

#### Certificate Object

> <!--

Reviewer: ```suggestion
<!-- YAML
```

changes:
  - version: REPLACEME
    pr-url: https://github.com/nodejs/node/pull/24358
    description: Support Elliptic Curve public key info.
-->

A certificate object has properties corresponding to the fields of the
certificate.

* `raw` {Buffer} The DER encoded X.509 certificate data.
* `subject` {Object} The certificate subject, described in terms of
   Country (`C:`), StateOrProvince (`ST`), Locality (`L`), Organization (`O`),
   OrganizationalUnit (`OU`), and CommonName (`CN`). The CommonName is typically
   a DNS name with TLS certificates. Example:
   `{C: 'UK', ST: 'BC', L: 'Metro', O: 'Node Fans', OU: 'Docs', CN: 'example.com'}`.
* `issuer` {Object} The certificate issuer, described in the same terms as the
   `subject`.
* `valid_from` {string} The date-time the certificate is valid from.
* `valid_to` {string} The date-time the certificate is valid to.
* `serialNumber` {string} The certificate serial number, as a hex string.

>    Example `'B9B0D332A1AA5635'`.

Reviewer: ```suggestion
   Example: `'B9B0D332A1AA5635'`.
```

* `fingerprint` {string} The SHA-1 digest of the DER encoded certificate. It is

>   returned as a `:` seperated hexadecimal string. Example: `'2A:7A:C2:DD:...'`.

Reviewer: ```suggestion
  returned as a `:` separated hexadecimal string. Example: `'2A:7A:C2:DD:...'`.
```

* `fingerprint256` {string} The SHA-256 digest of the DER encoded certificate.

>    It is returned as a `:` seperated hexdecimal string. Example:

Reviewer: ```suggestion
   It is returned as a `:` separated hexadecimal string. Example:
```

>    It is returned as a `:` separated hexdecimal string. Example:

Reviewer: ```suggestion
   It is returned as a `:` separated hexadecimal string. Example:
```

   `'2A:7A:C2:DD:...'`.
* `ext_key_usage` {Array} (Optional) The extended key usage, a set of OIDs.
* `subjectaltname` {Array} (Optional) An array of names for the subject, an
   alternative to the `subject` names.
* `infoAccess` {Array} (Optional) An array describing the AuthorityInfoAccess,
   used with OCSP.
* `issuerCert` {Object} (Optional) The issuer certificate object. For
   self-signed certificates, this may be a circular reference.

The certificate may contain information about the public key, depending on
the key type.

For RSA keys, the following properties may be defined:
* `bits` {number} The RSA bit size. Example: `1024`.
* `exponent` {string} The RSA exponent, as a string in hexadecimal number

>   notation.  Example `'0x010001'`

Reviewer: ```suggestion
  notation.  Example: `'0x010001'`.
```

   `'B56CE45CB7...'`.
* `pubkey` {Buffer} The RSA public key.

For EC keys, the following properties may be defined:
* `ecBits` {number} The bit-strength of the elliptic curve. Example: `256`.

> * `ecPubKey` {Buffer} The EC public key.

Reviewer 2: It's a minor inconsistency to camel-case `PubKey` when it's spelled `pubkey` for RSA keys. Perhaps `ecPubkey`?

Author: How about just `pubkey`? Then its the same as RSA. The cert property names are so wildly varying, from `infoAccess` to `ext_key_usage` to `subjectaltname` I figured rather than try any attempt at consistency, I'd just use our modern naming convention. I could do `ecPubkey`, but it feels a bit like the worst of both worlds. It isn't the same as RSA, but isn't camelCase either, it has the same RSA capitalization error.

Reviewer 2: `pubkey` WFM but makes it inconsistent with the other `ec*` properties. You could strip the ec prefix, I don't think that'd conflict with other properties, but makes it less clear they're EC-only properties. I'll just leave it up to you.

Author: I stripped ec, the "curve" is a pretty clear indication that its an elliptic curve. And I added `bits` for RSA for symmetricality. PTAL.
